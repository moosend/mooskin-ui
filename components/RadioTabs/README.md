# MooSkin UI - RadioTabs Component

The MooSkin RadioTabs Component is like a tabbed window, with as many radio tabs as you like.

___

### Usage

To start using the RadioTabs Component first you have to Import its two main components

```
Import {RadioTabs, Content} from 'mooskin';
```
or
```
import RadioTabs, {Content} from 'mooskin/lib/RadioTabs';
```

And then you can simply start using it by typing

```
<RadioTabs>
    <Content title="Title 1" active>
        <div style={{...style, backgroundColor: '#5CCDDF'}}>
            Content 1
        </div>
    </Content>
    <Content title="Title 2">
        <div style={{...style, backgroundColor: '#F48770'}}>
            Content 2
        </div>
    </Content>
    <Content title="Title 3">
        <div style={{...style, backgroundColor: '#F2C14A'}}>
            Content 3
        </div>
    </Content>
</RadioTabs>
```


Individual <Content> components has a required 'title' prop which will be displayed in the radio header, the rest are optional. You can put whatever inside <Content> tags, it will be displayed as content for that particular tab.

### Examples

Can be used without importing the <Content> component with shorthand dot(.) notation.

```
Import {RadioTabs} from 'mooskin';

<RadioTabs>
    <RadioTabs.Content title="Title 1" iconClass="fa fa-icon">
        <div style={{...style, backgroundColor: '#5CCDDF'}}>
            Content 1
        </div>
    </RadioTabs.Content>
    <RadioTabs.Content title="Title 2" materialIcon="home">
        <div style={{...style, backgroundColor: '#F48770'}}>
            Content 2
        </div>
    </RadioTabs.Content>
</RadioTabs>
```

an alternative view style can be toggled by passing the `horizontal` attribute to the <RadioTabs> component

```
Import {Content, RadioTabs} from 'mooskin';

<RadioTabs horizontal>
    <Content title="Title 1" iconClass="fa fa-icon">
        <div style={{...style, backgroundColor: '#5CCDDF'}}>
            Content 1
        </div>
    </Content>
    <Content title="Title 2" materialIcon="home">
        <div style={{...style, backgroundColor: '#F48770'}}>
            Content 2
        </div>
    </Content>
</RadioTabs>
```

## Supported attributes for ```<RadioTabs/>```

* `id` - id of the element
* `className` - css class
* `style` - extra styles for the main container
* `horizontal` - horizontal style of view for the RadioTabs


## Supported attributes for ```<Content/>```

* `title` - (required) sets the title for the radio tab
* `active` - sets which radio tab is active, the first is active by default
* `style` -  extra styles 

Allthough these attributes are supported, most of them are optional.

___

#### For more

[MooSkin-UI](https://github.com/moosend/mooskin-ui)
