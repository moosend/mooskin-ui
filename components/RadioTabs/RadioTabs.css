.containerHorizontal {
    display: grid;
    grid-gap: 10px;
}

.containerVertical {
    display: grid;
    grid-template-columns: 2fr 5fr;
    grid-gap: 10px;
}

.headerVertical {
    display: grid;
    grid-template-columns: 1fr;
    grid-gap: 10px;    
}

.headerHorizontal {
    display: grid;
    grid-auto-columns: 1fr;
    grid-gap: 10px;
}

.columnTemplateOne{
    grid-template-areas: "a";
}

.columnTemplateTwo{
    grid-template-areas: "a a";    
}

.columnTemplateInf{
    grid-template-areas: "a a a";    
}

.columnTemplateInfEven{
    grid-template-areas: "a a a a";    
}

.content {
    padding-left: 10px;
    color: #7b7b7b;
    font-size: 0.875rem;
    border-left: 1px solid #5ccdde;
}

.contentHorizontal{
    padding-top: 10px;
    border-top: 1px solid #5ccdde;
}

.radio span {
    margin-left: 10px;
    cursor: pointer;
    vertical-align: middle;
    user-select: none;
}

.radio {
    border-radius: 8px;
    color: #5ccdde;
    background: transparent;
    transition: 0.3s;
    border: 1px solid #5ccdde;
}

.radio:hover {
    background-color: #95dee9;
    color: #fff;
    border-color: #95dee9;
}

.radio input[type='radio'] {
    vertical-align: middle;
    appearance: none;
    margin: 0 auto;
    cursor: pointer;
    width: 1.188rem;
    height: 1.188rem;
    padding: 0.3425rem;
    background: #fff;
    border-radius:50%;
    border: 1px solid #5ccdde;
}

.radio:hover input[type='radio']{
    border-color: #95dee9;
}

.radio input[type='radio']:checked::after {
    background:#5ccdde;
    content:'';
    display:block;
    width: 0.375rem;
    height: 0.375rem;
    border-radius:50%; 
}

.radio label {
    display: inline-block;    
    cursor: pointer;
    padding: 12px;
    height: 100%;
    display: block; 
    border-radius: 8px;
    font-size: 0.875rem;
    width: 100%;
    height: 100%;
}

.checkedRadio {
    background: #5ccdde;
    color: #fff;
}

.checkedRadio:hover {
    background: #5ccdde;
    border-color: #5ccdde;
}

.checkedRadio:hover input[type='radio']{
    border-color: #5ccdde;
}

.visible {
    display: block;
}

.invisible {
    display: none;
}