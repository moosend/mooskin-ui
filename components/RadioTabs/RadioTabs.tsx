import * as React from 'react';

import styles from './RadioTabs.css';

// import {IRadioGroupProps, IRadioProps, Radio, RadioGroup} from '../Radio/Radio';

export interface IRadioTabsProps {

    /** override id */
    id?: string;

    /** container class */
    className?: string;

    /** override styles */
    style?: {[key: string]: string|number};

    /** horizontal RadioTabs view */
    horizontal?: boolean;

    /** children here can only be Tab elements */
    children?: Array<React.ReactElement<IContentProps>> | React.ReactElement<IContentProps>;
}

export interface IContentProps {

    /** container class */
    className?: string;

    /** title */
    title: string;

    /** if active */
    active?: boolean;

    /** override styles */
    style?: {[key: string]: string|number};

}

export interface IRadioTabsState {
    activeRadio?: number;
}

class RadioTabs extends React.Component<IRadioTabsProps, IRadioTabsState> {

    public static defaultProps = {
        // activeRadio: 0,
        className: '',
        style: {}
    };

    public static Content: React.StatelessComponent<IContentProps>;

    private name: string;

    constructor(props: IRadioTabsProps){
        super(props);

        this.name = this.generateName();

        this.state = {
            activeRadio: this.getActiveRadio()
        };
    }

    public render(){

        const {headers, bodies} = this.makeContent();
        const containerStyles = !this.props.horizontal ? styles.containerVertical : styles.containerHorizontal;
        const headerStyles = !this.props.horizontal ? styles.headerVertical : styles.headerHorizontal;
        const contentStyles = !this.props.horizontal ? styles.content : styles.contentHorizontal;
        const columnTemplate = !this.props.horizontal ? '' : this.columnTemplate(headers.length);

        return (
            <div className={`radio-tabs-component ${containerStyles}`}>
                <div className={`${headerStyles} ${columnTemplate}`}>
                    {headers}
                </div>
                <div className={contentStyles}>
                    {bodies}
                </div>
            </div>
        );
    }

    public onClickRadio = (radioIndex: number) => {
        return (e: React.MouseEvent<HTMLDivElement>) => {
            this.setState({activeRadio: radioIndex});
        };
    }

    private makeContent(){

        const headers: Array<React.ReactElement<IRadioTabsProps>> = [];
        const bodies: Array<React.ReactElement<IContentProps>> = [];

        React.Children.forEach(this.props.children, (child, index) => {
            if (React.isValidElement<IContentProps>(child)){

                headers.push(
                    <Header
                        key={index}
                        title={child.props.title}
                        active={this.state.activeRadio === index}
                        onClick={this.onClickRadio(index)}
                        name={this.name}
                    />
                );

                const extraProps: Partial<IContentProps & {key: number}> = {
                    active: this.state.activeRadio === index,
                    key: index,
                };

                bodies.push(React.cloneElement(child, extraProps));

            }else{
                throw new Error('<RadioTabs> element only accepts Content elements as children');
            }
        });

        return {headers, bodies};
    }

    private getActiveRadio(): number {
        let activeRadio = 0;
        const childrenArray = React.Children.toArray(this.props.children);

        for (const [index, value] of childrenArray.entries()){
            if (React.isValidElement<IContentProps>(value) && value.props.active){
                activeRadio = index;
            }
        }

        return activeRadio;
    }

    private columnTemplate = (headers: number) => {
        if (headers === 1) {
            return styles.columnTemplateOne;
        } else if (headers === 2){
            return styles.columnTemplateTwo;
        } else {
            if (headers === 6 || headers % 2 === 1){
                return styles.columnTemplateInf;
            } else {
                return styles.columnTemplateInfEven;
            }
        }

    }

    private generateName = () => {
        return Date.now().toString();
    }
}

export const Content: React.StatelessComponent<IContentProps> = (props) => {

    const displayClass = !props.active ? styles.invisible : styles.visible;

    return (
        <div
            className={`radio-content ${displayClass}`}
            style={props.style}
        >
            {props.children}
        </div>
    );
};

export interface IHeaderProps {
    title: string;
    active: boolean;
    name?: string;
    onClick: (e: React.MouseEvent<HTMLDivElement>) => void;
}

export const Header: React.StatelessComponent<IHeaderProps> = (props) => {

    const activeRadio = props.active ? styles.activeHeader : styles.inactiveHeader;
    const checkedRadio = !props.active ? '' : styles.checkedRadio;

    return (
        <div className={`radio-header ${styles.radio} ${activeRadio} ${checkedRadio}`} onClick={props.onClick}>
            <label>
                <input
                    type="radio"
                    name={props.name}
                    // onClick={onRadioClick}
                    // disabled={props.disabled}
                    defaultChecked={props.active}
                />
                <span>{props.title}</span>
            </label>
            {props.children}
        </div>
    );
};

export default RadioTabs;
