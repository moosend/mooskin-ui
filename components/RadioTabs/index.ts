import RadioTabs from './RadioTabs';

export default RadioTabs;
