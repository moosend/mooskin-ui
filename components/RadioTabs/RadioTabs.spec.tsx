import * as React from 'react';
import RadioTabs, {Content, Header} from './RadioTabs';

import { mount, render, shallow } from 'enzyme';

describe('RadioTabs', () => {

    test('renders RadioTabs properly according to snapshot', () => {

        Date.now = jest.fn(() => 1482363367071);

        const component = shallow(
            <RadioTabs id="5" className="mySwitch" style={{color: 'black'}} horizontal>
                <Content title="title1">Content1</Content>
                <Content title="title2" active>Content2</Content>
                <Content title="title3" className="classname">Content3</Content>
            </RadioTabs>
        );

        expect(component.find('Header').first().hasClass('columnTemplateInf')).toBeTruthy;
        expect(component.find('Header').length).toBe(3);
        expect(component.find('Content').length).toBe(3);

        expect(component.find('Content').first().prop('children')).toBe('Content1');

        expect(component).toMatchSnapshot();
    });

    test('renders Content properly according to snapshot', () => {

        const component = shallow(
            <Content
                title="title1"
                style={{color: 'blue'}}
                active
            >
                Some Content
            </Content>
        );

        expect(component).toMatchSnapshot();
    });

    test('renders Header properly according to snapshot', () => {

        Date.now = jest.fn(() => 1482363367071);

        const component = shallow(
            <Header
                title="title1"
                onClick={() => null}
                active
            />
        );

        expect(component).toMatchSnapshot();
    });

    test('changes the state and displays appropriate content when one of the headers is clicked and back', () => {

        const component = mount(
            <RadioTabs>
                <Content title="title1">Content1</Content>
                <Content title="title2" active>Content2</Content>
            </RadioTabs>
        );

        expect(component.state('activeRadio')).toBe(1);
        expect(component.find('.radio-header').first().find('input').prop('defaultChecked')).toBeFalsy();
        expect(component.find('.radio-header').last().find('input').prop('defaultChecked')).toBeTruthy();
        expect(component.find('.radio-content').first().hasClass('invisible')).toBeTruthy();
        expect(component.find('.radio-content').last().hasClass('visible')).toBeTruthy();

        component.find('.radio-header').first().simulate('click');

        expect(component.state('activeRadio')).toBe(0);
        expect(component.find('.radio-header').first().find('input').prop('defaultChecked')).toBeTruthy();
        expect(component.find('.radio-header').last().find('input').prop('defaultChecked')).toBeFalsy();
        expect(component.find('.radio-content').first().hasClass('visible')).toBeTruthy();
        expect(component.find('.radio-content').last().hasClass('invisible')).toBeTruthy();

        component.find('.radio-header').last().simulate('click');

        expect(component.state('activeRadio')).toBe(1);
        expect(component.find('.radio-header').first().find('input').prop('defaultChecked')).toBeFalsy();
        expect(component.find('.radio-header').last().find('input').prop('defaultChecked')).toBeTruthy();
        expect(component.find('.radio-content').first().hasClass('invisible')).toBeTruthy();
        expect(component.find('.radio-content').last().hasClass('visible')).toBeTruthy();

    });
});
