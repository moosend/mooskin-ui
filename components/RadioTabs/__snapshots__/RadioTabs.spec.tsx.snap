// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioTabs renders Content properly according to snapshot 1`] = `
<div
  className="radio-content visible"
  style={
    Object {
      "color": "blue",
    }
  }
>
  Some Content
</div>
`;

exports[`RadioTabs renders Header properly according to snapshot 1`] = `
<div
  className="radio-header radio activeHeader checkedRadio"
  onClick={[Function]}
>
  <label>
    <input
      defaultChecked={true}
      type="radio"
    />
    <span>
      title1
    </span>
  </label>
</div>
`;

exports[`RadioTabs renders RadioTabs properly according to snapshot 1`] = `
<div
  className="radio-tabs-component containerHorizontal"
>
  <div
    className="headerHorizontal columnTemplateInf"
  >
    <Header
      active={false}
      name="1482363367071"
      onClick={[Function]}
      title="title1"
    />
    <Header
      active={true}
      name="1482363367071"
      onClick={[Function]}
      title="title2"
    />
    <Header
      active={false}
      name="1482363367071"
      onClick={[Function]}
      title="title3"
    />
  </div>
  <div
    className="contentHorizontal"
  >
    <Content
      active={false}
      title="title1"
    >
      Content1
    </Content>
    <Content
      active={true}
      title="title2"
    >
      Content2
    </Content>
    <Content
      active={false}
      className="classname"
      title="title3"
    >
      Content3
    </Content>
  </div>
</div>
`;
